src/fixes/unordered_key.rs

```diff
@@ -0,0 +1,724 @@
+// control comments).
+impl Fix for UnorderedKeyFixer<'_> {
+    fn name(&self) -> &str {
+        self.name
+    }
+
+    fn fix_warnings(
+        &mut self,
+        warnings: Vec<&mut Warning>,
+        lines: &mut Vec<LineEntry>,
+    ) -> Option<usize> {
+        // We find all sorting groups and sort them
+
+        let mut start_index = 0;
+        let mut end = None;
+        let mut is_disabled = false;
+
+        for i in 0..lines.len() {
+            let line = &lines[i];
+
+            let mut is_control_comment = false;
+            let mut controls_this_check = false;
+            let mut is_off = false;
+
+            if let Some(comment) = line.get_control_comment() {
+                is_control_comment = true;
+                controls_this_check = comment.checks.contains(&self.name);
+                is_off = comment.is_disabled();
+            }
+
+            if !is_disabled {
+                if !line.is_empty_or_comment() {
+                    end.replace(i + 1);
+                }
+
+                let is_last = i == lines.len() - 1;
+                if line.is_empty() || is_last || is_control_comment {
+                    if let Some(end_index) = end {
+                        Self::sort_part(&mut lines[start_index..end_index]);
+                        end = None;
+                    }
+                    start_index = i + 1;
+                }
+            }
+
+            if controls_this_check {
+                is_disabled = is_off;
+                start_index = i + 1;
+            }
+        }
+
+        let count = warnings.len();
+        for warning in warnings {
+            warning.mark_as_fixed();
+        }
+
+        Some(count)
+    }
+}
+
+impl UnorderedKeyFixer<'_> {
+    fn sort_part(part: &mut [LineEntry]) {
+        // Each slice includes a significant line (with key) and previous comments (if present)
+        let mut slice_start = 0;
+        let mut slices = Vec::with_capacity(part.len());
+        for (i, line) in part.iter().enumerate() {
+            if !line.is_comment() {
+                slices.push(&part[slice_start..=i]);
+                slice_start = i + 1;
+            }
+        }
+
+        slices.sort_by_cached_key(|slice| {
+            // I think, that we should modify get_key() so it will return Option<&str> instead of
+            // Option<String>.
+            slice.last().expect("significant line").get_key()
+        });
+
+        let mut sorted_lines = Vec::with_capacity(part.len());
+        for slice in slices {
+            sorted_lines.extend_from_slice(slice);
+        }
+
+        part.clone_from_slice(sorted_lines.as_slice());
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::common::tests::line_entry;
+
+    fn get_lines(strs: Vec<&str>) -> Vec<LineEntry> {
+        let total = strs.len();
+
+        strs.iter()
+            .enumerate()
+            .map(|(i, &str)| line_entry(i + 1, total, str))
```

Comment:
```suggestion
            .map(|(i, &line)| line_entry(i + 1, total, line))
```
Please don't use `str` as a variable name, because Rust has the same type's name - [str](https://doc.rust-lang.org/std/primitive.str.html).

Reply:
Done